mythril/support/signatures.py

```diff
@@ -1,44 +1,244 @@
+#!/usr/bin/env python3
+# -*- coding: UTF-8 -*-
+"""mythril.py: Function Signature Database
+"""
 import re
+import os
+import json
+import time
+import pathlib
+import logging
 from ethereum import utils
 
-
-def add_signatures_from_file(file, sigs={}):
-
-    funcs = []
-
-    with open(file, encoding="utf-8") as f:
-
-        code = f.read()
-
-    funcs = re.findall(r'function[\s]+(\w+\([^\)]*\))', code, re.DOTALL)
-
-    for f in funcs:
-
-        f = re.sub(r'[\n]', '', f)
-
-        m = re.search(r'^([A-Za-z0-9_]+)', f)
-
-        if (m):
-
-            signature = m.group(1)
-
-            m = re.search(r'\((.*)\)', f)
-
-            _args = m.group(1).split(",")
-
-            types = []
-
-            for arg in _args:
-
-                _type = arg.lstrip().split(" ")[0]
-                if _type == "uint":
-                    _type = "uint256"
-
-                types.append(_type)
-
-            typelist = ",".join(types)
-            signature += "(" + typelist + ")"
-
-            signature = re.sub(r'\s', '', signature)
-
-            sigs["0x" + utils.sha3(signature)[:4].hex()] = signature
+# todo: tintinweb - make this a normal requirement? (deps: eth-abi and requests, both already required by mythril)
+try:
+    # load if available but do not fail
+    import ethereum_input_decoder
+    from ethereum_input_decoder.decoder import FourByteDirectoryOnlineLookupError
+except ImportError:
+    # fake it :)
+    ethereum_input_decoder = None
+    FourByteDirectoryOnlineLookupError = Exception
+
+
+class SimpleFileLock(object):
```

Comment:
hi @tintinweb , we can use some library here right?

Reply:
@norhh This is a quick shot on providing a solution for the concurrency issue until the switch is made to a real database (leveldb was mentioned earlier). The idea was to avoid adding another dependency. Free to propose something that works cross-platform.

Reply:
quicker shot would have been to use a library as it might anyway get replaced by the leveldb so not a much of longterm dependency 😄 anyway it doesn't matter now.

Reply:
In cases such as this I also would be cautious to add another dependency, it does not have all the same problems as a long term dependency might have.
One thing that could happen (specifically with short term dependencies) is us forgetting to remove the dependency when its not needed.